mini-projects/00-jq-rs/src/filters.rs

```diff
@@ -0,0 +1,278 @@
+use serde_json::Value;
+use std::fmt;
+use std::num::ParseIntError;
+use thiserror::Error;
+
+use crate::functions::{add_function, delete_function, length_function};
+
+#[derive(Error, Debug)]
+pub enum MyErrors {
+    #[error("Failed to read the provided JSON file: {0}")]
+    JSONError(#[from] serde_json::Error),
+
+    #[error("The specified key '{0}' was not found in the JSON data")]
+    KeyNotFound(String),
+
+    #[error("Invalid Needle: {0}")]
+    InvalidNeedle(String),
+
+    #[error("Dictionary not found")]
+    DictionaryNotFound,
+
+    #[error("List not found")]
+    ListNotFound,
+
+    #[error("You need an integer")]
+    ParseError(#[from] ParseIntError),
+
+    #[error("Index out of bounds")]
+    IndexOutOfBounds,
+
+    #[error("Missing brackets")]
+    MissingBrackets,
+
+    #[error("Invalid input for this function")]
+    InvalidInput,
+}
+
+fn object_identifier_filter(input: &Value, needle: &str) -> Result<Value, MyErrors> {
+    let key = &needle[1..];
+    if let Some(value) = input.get(key) {
+        Ok(value.clone())
+    } else {
+        Err(MyErrors::KeyNotFound(key.to_string()))
+    }
+}
+
+fn array_index(input: &Value, index_str: &str) -> Result<Value, MyErrors> {
+    let index: usize = index_str.parse().map_err(MyErrors::ParseError)?;
+
+    // Check bounds and return the value
+    if index < input.as_array().unwrap().len() {
+        Ok(input[index].clone())
+    } else {
+        Err(MyErrors::IndexOutOfBounds)
+    }
+}
+
+fn array_slice(input: &Value, index_str: &str) -> Result<Value, MyErrors> {
+    let indices: Vec<&str> = index_str.split(':').collect();
+    let start: usize = indices[0].parse().map_err(MyErrors::ParseError)?;
+    let end: usize = indices[1].parse().map_err(MyErrors::ParseError)?;
+
+    if start < end && end <= input.as_array().unwrap().len() {
+        let slice = &input.as_array().unwrap()[start..end];
+        Ok(Value::Array(slice.to_vec()))
+    } else {
+        Err(MyErrors::IndexOutOfBounds)
```

Comment:
this doesn't seem like the correct error in the first case of the if statement - start >= end is an incorrect bounds, not out of bounds